{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e6895853-7e46-4fd7-87b6-31b491061fe0",
   "metadata": {},
   "source": [
    "# [Uncalled4](https://www.nature.com/articles/s41592-025-02631-4); [GitHub](https://github.com/skovaka/uncalled4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4d36677d-6a42-4889-8296-7f8f6b1b3d7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: uncalled4 [subcommand] [-h] [-v]\n",
      "\n",
      "Utility for Nanopore Current ALignment to Large Expanses of DNA\n",
      "\n",
      "subcommand options:\n",
      "Signal Alignment:\n",
      "  align      Perform DTW alignment guided by basecalled alignments\n",
      "  train      Train new k-mer pore models\n",
      "  convert    Convert between signal alignment file formats\n",
      "\n",
      "Analysis:\n",
      "  refstats   Calculate per-reference-coordinate statistics\n",
      "  readstats  Compute per-read summary statistics\n",
      "  compare    Compare multiple alignments of the same read\n",
      "\n",
      "Visualization:\n",
      "  dotplot    Plot signal-to-reference alignment dotplots\n",
      "  trackplot  Plot alignment tracks and per-reference statistics\n",
      "  browser    Interactive signal alignment genome browser\n",
      "\n",
      "options:\n",
      "  -h, --help     show this help message and exit\n",
      "  -v, --version  show program's version number and exit\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "uncalled4=/v/o67ut7/.local/bin/uncalled4\n",
    "$uncalled4 -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "90be1b08-8bad-410e-b433-4a5826655048",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: uncalled4 align [-h] [--ref REF | --self] [--reads READS [READS ...]]\n",
      "                       --bam-in [BAM_IN] [-p PROCESSES] [--flowcell FLOWCELL]\n",
      "                       [--kit KIT] [--basecaller-profile BASECALLER_PROFILE]\n",
      "                       [--rna] [--ordered-out] [-f] [--kmer-shift KMER_SHIFT]\n",
      "                       [--bam-chunksize BAM_CHUNKSIZE] [--out-name OUT_NAME]\n",
      "                       [-r] [-l READ_FILTER] [-x READ_INDEX] [-n MAX_READS]\n",
      "                       [--count-events] [--del-max DEL_MAX]\n",
      "                       [--ins-max INS_MAX] [--unmask-splice] [--method METHOD]\n",
      "                       [-c {abs_diff,z_score,norm_pdf}] [-b BAND_WIDTH]\n",
      "                       [-s BAND_SHIFT] [--mvcmp-mask MVCMP_MASK]\n",
      "                       [--max-norm-dist MAX_NORM_DIST] [--max-sd MAX_SD]\n",
      "                       [--min-aln-length MIN_ALN_LENGTH]\n",
      "                       [-N {ref_mom,model_mom}] [--zero-ts] [-C CONFIG]\n",
      "                       [-o [BAM_OUT] | --tsv-out [TSV_OUT] | --eventalign-out\n",
      "                       [EVENTALIGN_OUT]] [-m PORE_MODEL] [--bam-f5c]\n",
      "                       [--tsv-cols TSV_COLS] [--tsv-na [TSV_NA]] [--tsv-noref]\n",
      "                       [--eventalign-flags EVENTALIGN_FLAGS]\n",
      "                       [--norm-iterations NORM_ITERATIONS]\n",
      "                       [--mask-skips [MASK_SKIPS]] [--skip-cost SKIP_COST]\n",
      "                       [--stay-cost STAY_COST] [--move-cost MOVE_COST]\n",
      "\n",
      "Perform DTW alignment guided by basecalled alignments\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "  --ref REF             Reference FASTA file, must match --bam-in reference\n",
      "                        (default: None)\n",
      "  --self                Perform reference-free signal-to-basecall alignment\n",
      "                        (default: False)\n",
      "  --reads READS [READS ...]\n",
      "                        Paths to FAST5, SLOW5, or POD5 files, or to\n",
      "                        directories containing those files (optionally\n",
      "                        recursive) (default: None)\n",
      "  --bam-in [BAM_IN]     BAM file used to guide signal alignment. Must contain\n",
      "                        'mv' and 'ts' basecaller tags. (default: None)\n",
      "  -p PROCESSES, --processes PROCESSES\n",
      "                        Number of parallel processes (default: 1)\n",
      "  --flowcell FLOWCELL   Flowcell used for sequencing (e.g. FLO-MIN106)\n",
      "                        (default: )\n",
      "  --kit KIT             Kit used for sequencing (e.g. SQK-LSK109) (default: )\n",
      "  --basecaller-profile BASECALLER_PROFILE\n",
      "                        Basecalling model profile name ('dna_r10.4.1_260bps',\n",
      "                        'dna_r10.4.1_400bps', 'dna_r9.4.1_400bps',\n",
      "                        'rna_r9.4.1_70bps', 'rna004_130bps') (default: None)\n",
      "  --rna                 RNA alignment (required for custom pore models)\n",
      "                        (default: None)\n",
      "  --ordered-out         Output alignments in the same order as BAM input\n",
      "                        (default if processes=1) (default: False)\n",
      "  -f, --overwrite       Overwrite existing tracks (default: False)\n",
      "  --kmer-shift KMER_SHIFT\n",
      "                        K-mer shift (default: None)\n",
      "  --bam-chunksize BAM_CHUNKSIZE\n",
      "                        Per-process alignment bam_chunksize (default: 500)\n",
      "  --out-name OUT_NAME   Output track name (defaults to file basename)\n",
      "                        (default: None)\n",
      "  -r, --recursive       Recursively search 'paths' for FAST5, SLOW5, or POD5\n",
      "                        files (default: False)\n",
      "  -l READ_FILTER, --read-filter READ_FILTER\n",
      "                        List of read IDs to load, or file containing one read\n",
      "                        ID per line (default: None)\n",
      "  -x READ_INDEX, --read-index READ_INDEX\n",
      "                        File containing a mapping of read IDs to filenames\n",
      "                        (default: None)\n",
      "  -n MAX_READS, --max-reads MAX_READS\n",
      "                        Maximum number of reads to load (default: None)\n",
      "  --count-events        Compute and output per-reference event counts, where N\n",
      "                        > 1 indicates N-1 stays and N < 1 indicates 1/N skips\n",
      "                        (default: False)\n",
      "  --del-max DEL_MAX     Will remove reference positions overlapping deletions\n",
      "                        longer than this (default: 10)\n",
      "  --ins-max INS_MAX     Will remove events overlapping insertions longer than\n",
      "                        this (default: 10)\n",
      "  --unmask-splice       Save DTW band coordinates to database (default: False)\n",
      "  --method METHOD       Alignment method (default=\"bcdtw\", or set to \"moves\"\n",
      "                        to use unprocessed ref-moves) (default: bcdtw)\n",
      "  -c {abs_diff,z_score,norm_pdf}, --cost-fn {abs_diff,z_score,norm_pdf}\n",
      "                        DTW cost function (default: abs_diff)\n",
      "  -b BAND_WIDTH, --band-width BAND_WIDTH\n",
      "                        DTW band width (default: 50)\n",
      "  -s BAND_SHIFT, --band-shift BAND_SHIFT\n",
      "                        DTW band shift (default: 0.5)\n",
      "  --mvcmp-mask MVCMP_MASK\n",
      "                        Will filter out positions with mvcmp.dist >=\n",
      "                        mvcmp_mask if specified (default: None)\n",
      "  --max-norm-dist MAX_NORM_DIST\n",
      "                        Maximum mvcmp.dist for posititions to be used for\n",
      "                        iterative normalization (default: 2.0)\n",
      "  --max-sd MAX_SD       Maximum current standard deviation (default: 1.0)\n",
      "  --min-aln-length MIN_ALN_LENGTH\n",
      "                        Minimum number of aligned bases (default: 100)\n",
      "  -N {ref_mom,model_mom}, --norm-mode {ref_mom,model_mom}\n",
      "                        Normalization mode (ref_mom, model_mom) (default:\n",
      "                        ref_mom)\n",
      "  --zero-ts             Set 'ts' BAM tag to 0 (Dorado 0.5.3 bug workaround)\n",
      "                        (default: False)\n",
      "  -C CONFIG, --config CONFIG\n",
      "                        Configuration file in TOML format (default: None)\n",
      "  -o [BAM_OUT], --bam-out [BAM_OUT]\n",
      "                        BAM output file (default: None)\n",
      "  --tsv-out [TSV_OUT]   TSV output file (or \"-\"/no argument for stdout)\n",
      "                        (default: None)\n",
      "  --eventalign-out [EVENTALIGN_OUT]\n",
      "                        Eventalign (nanopolish) output file (or \"-\"/no\n",
      "                        argument for stdout) (default: None)\n",
      "  -m PORE_MODEL, --pore-model PORE_MODEL\n",
      "                        Model preset name or TSV filename (default: None)\n",
      "  --bam-f5c             Include f5c ss: and si: BAM output tags (default:\n",
      "                        False)\n",
      "  --tsv-cols TSV_COLS   TSV file output alignment layers (comma-separated, can\n",
      "                        also include \"read_id\" (default: dtw)\n",
      "  --tsv-na [TSV_NA]     Missing value representation for TSV output (default:\n",
      "                        *)\n",
      "  --tsv-noref           Will NOT output reference coordinates to TSV if True\n",
      "                        (default: False)\n",
      "  --eventalign-flags EVENTALIGN_FLAGS\n",
      "                        Eventalign optional flags (comma-separated list of\n",
      "                        \"\"print-read-names\", \"signal-index\", \"samples\")\n",
      "                        (default: [])\n",
      "  --norm-iterations NORM_ITERATIONS\n",
      "                        Number of DTW iterations to perform (default: 2)\n",
      "  --mask-skips [MASK_SKIPS]\n",
      "                        Either \"all\" to mask all skips, or \"keep_best\" to mask\n",
      "                        all but the closest to the model (default: None)\n",
      "  --skip-cost SKIP_COST\n",
      "                        DTW event skip (vertical) penalty (default: 2.0)\n",
      "  --stay-cost STAY_COST\n",
      "                        DTW event stay (horizontal) penalty (default: 1.0)\n",
      "  --move-cost MOVE_COST\n",
      "                        DTW event move (diagonal) penalty (default: 1.0)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "FAST5='/v/projects/nanopore/agnes/signal_visualization/data/timin0421/workspace/'\n",
    "REF=\"/v/volumes/nanopore/ref/p300_ID3.fasta\"\n",
    "BAM='/v/projects/nanopore/agnes/signal_visualization/data/timin0421/pass/bam_runid_db4f2d165a2cb95d870762f8bd1749edb5a64d5d_0_0.bam'\n",
    "OUT_NAME='/v/projects/nanopore/balazs/data/uncalled_data/'\n",
    "\n",
    "uncalled4=/v/o67ut7/.local/bin/uncalled4\n",
    "$uncalled4 align -h\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0d92cbf1-9cbe-48c5-b75f-69d598eb6392",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading read index '/v/projects/nanopore/agnes/signal_visualization/data/timin0421/workspace/read_index.txt'\n",
      "Auto-detected flowcell='FLO-FLG001' kit='SQK-LSK110'\n",
      "Using model 'dna_r9.4.1_400bps_6mer'\n",
      "Counter()\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "FAST5='/v/projects/nanopore/agnes/signal_visualization/data/timin0421/workspace/'\n",
    "REF=\"/v/volumes/nanopore/ref/p300_ID3.fasta\"\n",
    "BAM='/v/projects/nanopore/agnes/signal_visualization/data/timin0421/pass/basecalls.bam'\n",
    "OUT_PATH='/v/projects/nanopore/balazs/data/uncalled_data/'\n",
    "\n",
    "uncalled4=/v/o67ut7/.local/bin/uncalled4\n",
    "$uncalled4 align --ref $REF --reads $FAST5  --bam-in $BAM -r  --tsv-out $OUT_PATH/out.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6eeda876-55b4-4a84-ad8b-3f84fa1b7a65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: uncalled4 align [-h] [--ref REF | --self] [--reads READS [READS ...]]\n",
      "                       --bam-in [BAM_IN] [-p PROCESSES] [--flowcell FLOWCELL]\n",
      "                       [--kit KIT] [--basecaller-profile BASECALLER_PROFILE]\n",
      "                       [--rna] [--ordered-out] [-f] [--kmer-shift KMER_SHIFT]\n",
      "                       [--bam-chunksize BAM_CHUNKSIZE] [--out-name OUT_NAME]\n",
      "                       [-r] [-l READ_FILTER] [-x READ_INDEX] [-n MAX_READS]\n",
      "                       [--count-events] [--del-max DEL_MAX]\n",
      "                       [--ins-max INS_MAX] [--unmask-splice] [--method METHOD]\n",
      "                       [-c {abs_diff,z_score,norm_pdf}] [-b BAND_WIDTH]\n",
      "                       [-s BAND_SHIFT] [--mvcmp-mask MVCMP_MASK]\n",
      "                       [--max-norm-dist MAX_NORM_DIST] [--max-sd MAX_SD]\n",
      "                       [--min-aln-length MIN_ALN_LENGTH]\n",
      "                       [-N {ref_mom,model_mom}] [--zero-ts] [-C CONFIG]\n",
      "                       [-o [BAM_OUT] | --tsv-out [TSV_OUT] | --eventalign-out\n",
      "                       [EVENTALIGN_OUT]] [-m PORE_MODEL] [--bam-f5c]\n",
      "                       [--tsv-cols TSV_COLS] [--tsv-na [TSV_NA]] [--tsv-noref]\n",
      "                       [--eventalign-flags EVENTALIGN_FLAGS]\n",
      "                       [--norm-iterations NORM_ITERATIONS]\n",
      "                       [--mask-skips [MASK_SKIPS]] [--skip-cost SKIP_COST]\n",
      "                       [--stay-cost STAY_COST] [--move-cost MOVE_COST]\n",
      "\n",
      "Perform DTW alignment guided by basecalled alignments\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "  --ref REF             Reference FASTA file, must match --bam-in reference\n",
      "                        (default: None)\n",
      "  --self                Perform reference-free signal-to-basecall alignment\n",
      "                        (default: False)\n",
      "  --reads READS [READS ...]\n",
      "                        Paths to FAST5, SLOW5, or POD5 files, or to\n",
      "                        directories containing those files (optionally\n",
      "                        recursive) (default: None)\n",
      "  --bam-in [BAM_IN]     BAM file used to guide signal alignment. Must contain\n",
      "                        'mv' and 'ts' basecaller tags. (default: None)\n",
      "  -p PROCESSES, --processes PROCESSES\n",
      "                        Number of parallel processes (default: 1)\n",
      "  --flowcell FLOWCELL   Flowcell used for sequencing (e.g. FLO-MIN106)\n",
      "                        (default: )\n",
      "  --kit KIT             Kit used for sequencing (e.g. SQK-LSK109) (default: )\n",
      "  --basecaller-profile BASECALLER_PROFILE\n",
      "                        Basecalling model profile name ('dna_r10.4.1_260bps',\n",
      "                        'dna_r10.4.1_400bps', 'dna_r9.4.1_400bps',\n",
      "                        'rna_r9.4.1_70bps', 'rna004_130bps') (default: None)\n",
      "  --rna                 RNA alignment (required for custom pore models)\n",
      "                        (default: None)\n",
      "  --ordered-out         Output alignments in the same order as BAM input\n",
      "                        (default if processes=1) (default: False)\n",
      "  -f, --overwrite       Overwrite existing tracks (default: False)\n",
      "  --kmer-shift KMER_SHIFT\n",
      "                        K-mer shift (default: None)\n",
      "  --bam-chunksize BAM_CHUNKSIZE\n",
      "                        Per-process alignment bam_chunksize (default: 500)\n",
      "  --out-name OUT_NAME   Output track name (defaults to file basename)\n",
      "                        (default: None)\n",
      "  -r, --recursive       Recursively search 'paths' for FAST5, SLOW5, or POD5\n",
      "                        files (default: False)\n",
      "  -l READ_FILTER, --read-filter READ_FILTER\n",
      "                        List of read IDs to load, or file containing one read\n",
      "                        ID per line (default: None)\n",
      "  -x READ_INDEX, --read-index READ_INDEX\n",
      "                        File containing a mapping of read IDs to filenames\n",
      "                        (default: None)\n",
      "  -n MAX_READS, --max-reads MAX_READS\n",
      "                        Maximum number of reads to load (default: None)\n",
      "  --count-events        Compute and output per-reference event counts, where N\n",
      "                        > 1 indicates N-1 stays and N < 1 indicates 1/N skips\n",
      "                        (default: False)\n",
      "  --del-max DEL_MAX     Will remove reference positions overlapping deletions\n",
      "                        longer than this (default: 10)\n",
      "  --ins-max INS_MAX     Will remove events overlapping insertions longer than\n",
      "                        this (default: 10)\n",
      "  --unmask-splice       Save DTW band coordinates to database (default: False)\n",
      "  --method METHOD       Alignment method (default=\"bcdtw\", or set to \"moves\"\n",
      "                        to use unprocessed ref-moves) (default: bcdtw)\n",
      "  -c {abs_diff,z_score,norm_pdf}, --cost-fn {abs_diff,z_score,norm_pdf}\n",
      "                        DTW cost function (default: abs_diff)\n",
      "  -b BAND_WIDTH, --band-width BAND_WIDTH\n",
      "                        DTW band width (default: 50)\n",
      "  -s BAND_SHIFT, --band-shift BAND_SHIFT\n",
      "                        DTW band shift (default: 0.5)\n",
      "  --mvcmp-mask MVCMP_MASK\n",
      "                        Will filter out positions with mvcmp.dist >=\n",
      "                        mvcmp_mask if specified (default: None)\n",
      "  --max-norm-dist MAX_NORM_DIST\n",
      "                        Maximum mvcmp.dist for posititions to be used for\n",
      "                        iterative normalization (default: 2.0)\n",
      "  --max-sd MAX_SD       Maximum current standard deviation (default: 1.0)\n",
      "  --min-aln-length MIN_ALN_LENGTH\n",
      "                        Minimum number of aligned bases (default: 100)\n",
      "  -N {ref_mom,model_mom}, --norm-mode {ref_mom,model_mom}\n",
      "                        Normalization mode (ref_mom, model_mom) (default:\n",
      "                        ref_mom)\n",
      "  --zero-ts             Set 'ts' BAM tag to 0 (Dorado 0.5.3 bug workaround)\n",
      "                        (default: False)\n",
      "  -C CONFIG, --config CONFIG\n",
      "                        Configuration file in TOML format (default: None)\n",
      "  -o [BAM_OUT], --bam-out [BAM_OUT]\n",
      "                        BAM output file (default: None)\n",
      "  --tsv-out [TSV_OUT]   TSV output file (or \"-\"/no argument for stdout)\n",
      "                        (default: None)\n",
      "  --eventalign-out [EVENTALIGN_OUT]\n",
      "                        Eventalign (nanopolish) output file (or \"-\"/no\n",
      "                        argument for stdout) (default: None)\n",
      "  -m PORE_MODEL, --pore-model PORE_MODEL\n",
      "                        Model preset name or TSV filename (default: None)\n",
      "  --bam-f5c             Include f5c ss: and si: BAM output tags (default:\n",
      "                        False)\n",
      "  --tsv-cols TSV_COLS   TSV file output alignment layers (comma-separated, can\n",
      "                        also include \"read_id\" (default: dtw)\n",
      "  --tsv-na [TSV_NA]     Missing value representation for TSV output (default:\n",
      "                        *)\n",
      "  --tsv-noref           Will NOT output reference coordinates to TSV if True\n",
      "                        (default: False)\n",
      "  --eventalign-flags EVENTALIGN_FLAGS\n",
      "                        Eventalign optional flags (comma-separated list of\n",
      "                        \"\"print-read-names\", \"signal-index\", \"samples\")\n",
      "                        (default: [])\n",
      "  --norm-iterations NORM_ITERATIONS\n",
      "                        Number of DTW iterations to perform (default: 2)\n",
      "  --mask-skips [MASK_SKIPS]\n",
      "                        Either \"all\" to mask all skips, or \"keep_best\" to mask\n",
      "                        all but the closest to the model (default: None)\n",
      "  --skip-cost SKIP_COST\n",
      "                        DTW event skip (vertical) penalty (default: 2.0)\n",
      "  --stay-cost STAY_COST\n",
      "                        DTW event stay (horizontal) penalty (default: 1.0)\n",
      "  --move-cost MOVE_COST\n",
      "                        DTW event move (diagonal) penalty (default: 1.0)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "uncalled4=/v/o67ut7/.local/bin/uncalled4\n",
    "$uncalled4 align -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68f57b8e-3c1c-4fdb-a817-e55fae261d7c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "75669055-6bc0-467b-bbeb-2773e77d170c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "652e8c17-53a5-40be-9200-700a1bc9c3c9",
   "metadata": {},
   "outputs": [
    {
     "ename": "EmptyDataError",
     "evalue": "No columns to parse from file",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mEmptyDataError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[17], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/v/projects/nanopore/balazs/data/uncalled_data/out.tsv\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      2\u001b[0m df\u001b[38;5;241m.\u001b[39mhead()\n",
      "File \u001b[0;32m/opt/conda/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1026\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1013\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m   1014\u001b[0m     dialect,\n\u001b[1;32m   1015\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1022\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m   1023\u001b[0m )\n\u001b[1;32m   1024\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m-> 1026\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m _read(filepath_or_buffer, kwds)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.12/site-packages/pandas/io/parsers/readers.py:620\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    617\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    619\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 620\u001b[0m parser \u001b[38;5;241m=\u001b[39m TextFileReader(filepath_or_buffer, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwds)\n\u001b[1;32m    622\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    623\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m/opt/conda/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1620\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1617\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1619\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1620\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_make_engine(f, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mengine)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1898\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1895\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg)\n\u001b[1;32m   1897\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1898\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m mapping[engine](f, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions)\n\u001b[1;32m   1899\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m   1900\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/opt/conda/lib/python3.12/site-packages/pandas/io/parsers/c_parser_wrapper.py:93\u001b[0m, in \u001b[0;36mCParserWrapper.__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m     90\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype_backend\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpyarrow\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m     91\u001b[0m     \u001b[38;5;66;03m# Fail here loudly instead of in cython after reading\u001b[39;00m\n\u001b[1;32m     92\u001b[0m     import_optional_dependency(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpyarrow\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 93\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reader \u001b[38;5;241m=\u001b[39m parsers\u001b[38;5;241m.\u001b[39mTextReader(src, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwds)\n\u001b[1;32m     95\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39munnamed_cols \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reader\u001b[38;5;241m.\u001b[39munnamed_cols\n\u001b[1;32m     97\u001b[0m \u001b[38;5;66;03m# error: Cannot determine type of 'names'\u001b[39;00m\n",
      "File \u001b[0;32mparsers.pyx:581\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mEmptyDataError\u001b[0m: No columns to parse from file"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('/v/projects/nanopore/balazs/data/uncalled_data/out.tsv')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e146c3a-3cf1-4b59-9c2a-769fcaef093b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
